File: days/day_18/main.py
import re
import time
import os


def eval_plus_first(expression):
	while '(' in expression:
		expression = re.sub(r'\(([0-9+* ]+)\)', 
			lambda match: str(eval_plus_first(match.group(1))), 
			expression)

	while '+' in expression:
		expression = re.sub(r'([0-9]+) (\+) ([0-9]+)', eval_operation, expression, 1)
	
	while '*' in expression:
		expression = re.sub(r'([0-9]+) (\*) ([0-9]+)', eval_operation, expression, 1)

	return int(expression)


def eval_left_to_right(expression):
	while '(' in expression:
		expression = re.sub(r'\(([0-9+* ]+)\)', 
			lambda match: str(eval_left_to_right(match.group(1))), 
			expression)

	while '+' in expression or '*' in expression:
		expression = re.sub(r'^([0-9]+) ([+*]) ([0-9]+)', 
			eval_operation, 
			expression)

	return int(expression)


def eval_operation(match):
	x = int(match.group(1))
	y = int(match.group(3))
	oper = match.group(2)
	
	if oper == '+':
		return str(x+y)
	elif oper == '*':
		return str(x*y)
	else:
		return '0'


def part_one(expressions):
	total = 0
	for expression in expressions:
		total += eval_left_to_right(expression)
	return total


def part_two(expressions):
	total = 0
	for expression in expressions:
		total += eval_plus_first(expression)
	return total


def main():
	start_time = time.time()
	
	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		expressions = [line.strip() for line in data.readlines()]

		part_one_ans = part_one(expressions)
		part_two_ans = part_two(expressions)

		print('Day 18 ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))


if __name__ == "__main__":
		main()

File: days/day_18/main.js
const fs = require('fs');
const readline = require('readline');

async function loadInput() {
  const rl = readline.createInterface({
    input: fs.createReadStream(__dirname + '/input.txt')
  });

  const lines = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines;
}

function evalLeftToRight(expression) {
  while (expression.includes('(')) {
    expression = expression.replace(/\(([0-9+* ]+)\)/, 
      (match, p1) => String(evalLeftToRight(p1)));
  }

  while (expression.includes('+') || expression.includes('*')) {
    expression = expression.replace(/^([0-9]+) ([+*]) ([0-9]+)/, 
      (match, p1, p2, p3) => eval_operation(p1, p2, p3));
  }

  return Number(expression);
}

function evalPlusFirst(expression) {
  while (expression.includes('(')) {
    expression = expression.replace(/\(([0-9+* ]+)\)/, 
      (match, p1) => String(evalPlusFirst(p1)));
  }

  while (expression.includes('+')) {
    expression = expression.replace(/([0-9]+) (\+) ([0-9]+)/, 
      (match, p1, p2, p3) => eval_operation(p1, p2, p3));
  }

  while (expression.includes('*')) {
    expression = expression.replace(/([0-9]+) (\*) ([0-9]+)/, 
      (match, p1, p2, p3) => eval_operation(p1, p2, p3));
  }

  return Number(expression);
}


function eval_operation(x, oper, y) {
  if (oper == '+') {
    return String(Number(x) + Number(y));
  }
  else if (oper == '*') {
    return String(Number(x) * Number(y));
  }
  else {
    throw 'nope';
  }
}

function partOne(expressions) {
  return expressions.reduce((total, expression) => total + evalLeftToRight(expression), 0);
}

function partTwo(expressions) {
  return expressions.reduce((total, expression) => total + evalPlusFirst(expression), 0);
}

async function main() {
	const start_time = process.hrtime();
  
  const expressions = await loadInput();
  
  const part_one_ans = partOne(expressions);
  const part_two_ans = partTwo(expressions);
  
  const diff = process.hrtime(start_time);
  const total_time = (diff[0] + diff[1]/1e9).toFixed(3);
  
  console.log(`Day 18 (${total_time}s)`);
  console.log(`  Part 1: ${part_one_ans}`);
  console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
  main();
}